cmd/snap/cmd_model.go

```diff
@@ -0,0 +1,229 @@
+assertion.
+
+The --verbose option prints off more headers from the assertion.
+
+The --assertion option prints off the raw assertion.
+`)
+
+	// this list is a "nice" "human" "readable" "ordering" of headers to print
+	// off, sorted in lexographical order with meta headers and primary key
+	// headers removed, and big nasty keys such as device-key-sha3-384 and
+	// device-key at the bottom
+	// it also contains both serial and model assertion headers, but we
+	// follow the same code path for both assertion types and some of the
+	// headers are shared between the two, so it still works out correctly
+	niceOrdering = []string{
+		"architecture",
+		"base",
+		"classic",
+		"display-name",
+		"gadget",
+		"kernel",
+		"revision",
+		"timestamp",
+		"required-snaps",
+		"device-key-sha3-384",
+		"device-key",
+	}
+)
+
+type cmdModel struct {
+	waitMixin
+	timeMixin
+	Serial    bool `long:"serial"`
+	Verbose   bool `long:"verbose"`
+	Assertion bool `long:"assertion"`
+}
+
+func init() {
+	addCommand("model",
+		shortModelHelp,
+		longModelHelp,
+		func() flags.Commander {
+			return &cmdModel{}
+		}, timeDescs.also(waitDescs.also(map[string]string{
+			"assertion": i18n.G("Print the raw assertion."),
+			"verbose":   i18n.G("Print more of the assertion fields."),
+			"serial":    i18n.G("Print the serial assertion instead of the model."),
+		})),
+		[]argDesc{},
+	)
+}
+
+func (x *cmdModel) Execute(args []string) error {
+	var assertion asserts.Assertion
+	var err error
+	if x.Serial {
+		assertion, err = x.client.CurrentSerialAssertion()
+	} else {
+		assertion, err = x.client.CurrentModelAssertion()
+	}
+
+	if err != nil && err.Error() == client.ErrorKindAssertionNotFound {
+		// device is not registered yet - print helpful message to stdout and
+		// return err to set exit code non-zero
+		fmt.Fprintln(Stdout, "device not ready - no assertion found")
+		return err
+	}
+
+	if err != nil {
+		return err
+	}
+
+	// --assertion means output the raw assertion
+	if x.Assertion {
+		_, err = Stdout.Write(asserts.Encode(assertion))
+		return err
+	}
+
+	termWidth, _ := termSize()
+	termWidth -= 3
+	if termWidth > 100 {
+		// any wider than this and it gets hard to read
+		termWidth = 100
+	}
+
+	w := tabWriter()
+
+	// output the primary keys first in their canonical order
+	for _, headerName := range assertion.Type().PrimaryKey {
+		fmt.Fprintf(w, "%s:\t%s\n", headerName, assertion.HeaderString(headerName))
+	}
+
+	// --verbose means output all of the fields
+	if x.Verbose {
+		allHeadersMap := assertion.Headers()
+
+		for _, headerName := range niceOrdering {
+			headerValue, ok := allHeadersMap[headerName]
+			// make sure the header is in the map
+			if !ok {
+				continue
+			}
+
+			// switch on which header it is to handle some special cases
+			switch headerName {
+			// scalar values
+			case "architecture", "base", "classic", "display-name", "gadget", "kernel", "revision", "store":
```

Comment:
this could just be the default stanza, instead of listing these again

Reply:
The reason I went with listing all of these here was so that we didn't inadvertently start printing a header key which we haven't vetted to print nicely and so if there was a new key not listed here in a model assertion, it may be printed in a not-nicely formatted way, or worse it may even not be a scalar string in which case the type assertion may fail. As such I think it's better to only support specific keys for now here.

Reply:
but we are not iterating over all the headers, we are iterating only over local niceOrdering no?

Reply:
Actually yes you're right, sorry I was confused